xunit-autolabeler-v2/ast_parser/core/polyglot_parser.py

```diff
@@ -0,0 +1,188 @@
+
+
+def get_region_tag_regions(
+    source_path: str
+) -> Tuple[List[Tuple[str, int, int]], List[str]]:
+    """Get the region tag data from a given file (of any language)
+
+    Args:
+        source_path: path to the target file
+
+    Returns:
+        A tuple of the form (regions_and_tags, ignored_tag_names), where:
+          - regions_and_tags: a list of tuples of the form
+                            (region tag value, start line, end line)
+          - ignored_tag_names: a list of region tags (as strings) ignored
+                               due to cross-parser constants
+    """
+    def _get_region_tag_from_line(line: str) -> Tuple[int, str]:
+        """Extract line numbers and region tags
+           from a given (line number, text) 2-tuple.
+
+        Args:
+            line: a 2-tuple of the form (1-indexed line number, line text)
+
+        Returns:
+            a 2-tuple of the form (0-indexed line number, region tag)
+        """
+        (line_num, line_text) = line
+
+        tag = line_text.split(' ')[-1]
+        tag = tag.split(']')[0]
+        return (line_num + 1, tag)  # +1 = convert to 0-indexed
+
+    with open(source_path, 'r') as file:
+        content_lines = [(idx, ln) for idx, ln in enumerate(file.readlines())]
+
+        start_tag_lines = [t for t in content_lines if ' [START' in t[1]]
+        end_tag_lines = [t for t in content_lines if ' [END' in t[1]]
+
+        # region tags can be repeated, so we can't use them as dict keys
+        # for specific region blocks - so we use tuple arrays instead
+        start_regions = [_get_region_tag_from_line(t) for t in start_tag_lines]
+        end_regions = [_get_region_tag_from_line(t) for t in end_tag_lines]
+
+        unique_tag_names = list(set([region[1] for region in start_regions]))
+
+        # ignore "useless" region tags
+        ignored_tag_names = [tag for tag in unique_tag_names if
+                             tag in IGNORED_REGION_TAGS]
+        unique_tag_names = [tag for tag in unique_tag_names if
+                            tag not in ignored_tag_names]
+
+        if len(start_regions) != len(end_regions):
+            raise ValueError('Mismatched region tags: ' + source_path)
+
+        start_regions.sort()
+        end_regions.sort()
+
+        regions_and_tags = []
+        for tag in unique_tag_names:
+            matching_starts = [region for region in start_regions
+                               if region[1] == tag]
+            matching_ends = [region for region in end_regions
+                             if region[1] == tag]
+
+            if len(matching_starts) != len(matching_ends):
+                raise ValueError(
+                    f'Mismatched region tag [{tag}] in {source_path}')
+
+            for i in range(len(matching_starts)):
+                start_region = matching_starts[i]
+                regions_and_tags.append(
+                    (start_region[1], start_region[0], matching_ends[i][0])
+                )
```

Comment:
I might be lost in the code at this point, but I'm curious why are you using range(len(list)) instead of just looping the list with `for item in list`? Do you really need the index of the list in this block?

Reply:
Yes, because I'm merging multiple arrays (`matching_starts` and `matching_ends`) here.

Reply:
Oh I see, then it's OK. zip could possibly help in this type of operation.